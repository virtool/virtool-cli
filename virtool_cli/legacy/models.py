"""Notes
-----
* `name`
  * Must be unique across all OTUs in the repo.
* `_id`.
  * Must have a unique across all other OTUs in the repo.
* `isolates`
  * Must have only one isolate with `default` set to `true`.

"""

from enum import Enum
from typing import Annotated

from attr import dataclass
from pydantic import (
    BaseModel,
    Field,
    StringConstraints,
    field_validator,
    model_validator,
)


def transform_to_uppercase(string: str) -> str:
    return string.upper()


class LegacySourceType(str, Enum):
    CLONE = "clone"
    CULTURE = "culture"
    ISOLATE = "isolate"
    GENBANK = "genbank"
    GENOTYPE = "genotype"
    SEROTYPE = "serotype"
    STRAIN = "strain"
    VARIANT = "variant"


@dataclass
class LegacyIsolateSource:
    name: str
    type: LegacySourceType


class LegacySequence(BaseModel):
    id: str = Field(alias="_id")
    accession: Annotated[
        str,
        StringConstraints(
            min_length=4,
            pattern=r"^\w+\.\d+$",
            strip_whitespace=True,
            to_upper=True,
        ),
    ]
    definition: Annotated[str, StringConstraints(min_length=10, strip_whitespace=True)]
    host: Annotated[str, StringConstraints(strip_whitespace=True)]
    segment: str
    sequence: Annotated[
        str,
        StringConstraints(
            min_length=10,
            pattern="^[ATCGRYKMSWBDHVNatcgrykmswbdhvn]+$",
            strip_whitespace=True,
            to_upper=True,
        ),
    ]


class LegacyIsolate(BaseModel):
    id: str
    default: bool
    sequences: Annotated[list[LegacySequence], StringConstraints(min_length=1)]
    source_type: LegacySourceType
    source_name: Annotated[str, StringConstraints(min_length=1, strip_whitespace=True)]


class LegacySchemaSegment(BaseModel):
    molecule: str
    name: str
    required: bool


class LegacyOTU(BaseModel):
    id: Annotated[str, Field(alias="_id")]
    abbreviation: Annotated[str, StringConstraints(strip_whitespace=True)]
    isolates: Annotated[list[LegacyIsolate], Field(min_length=1)]
    name: Annotated[str, StringConstraints(min_length=5, strip_whitespace=True)]
    otu_schema: Annotated[
        list[LegacySchemaSegment],
        Field(min_length=1, alias="schema"),
    ]
    taxid: int

    @field_validator("isolates")
    @classmethod
    def check_default_isolate(cls, v):
        """Check if there is only one default isolate."""
        default_isolates = [isolate for isolate in v if isolate.default]

        if len(default_isolates) == 0:
            raise ValueError("At least one isolate must be default")

        if len(default_isolates) > 1:
            raise ValueError("Only one isolate can be default")

        return v

    @field_validator("otu_schema")
    @classmethod
    def check_schema_molecule(cls, v):
        """Check if all segments in the schema have the same molecule."""
        molecules = {segment.molecule for segment in v}

        if len(molecules) > 1:
            raise ValueError("All segments in a schema must have the same molecule")

        return v

    @field_validator("otu_schema")
    @classmethod
    def check_schema_name(cls, v):
        """Check if there are duplicate schema segment names."""
        names = {segment.name for segment in v}

        if len(names) != len(v):
            raise ValueError("All schema segments must have a unique name")

        return v

    @model_validator(mode="after")
    def check_schema_and_sequences(self) -> "LegacyOTU":
        """Make sure all sequences in the OTU have a corresponding schema segment."""
        schema_segments = {segment.name for segment in self.otu_schema}

        required_schema_segments = {
            segment.name for segment in self.otu_schema if segment.required
        }

        for isolate in self.isolates:
            for sequence in isolate.sequences:
                if sequence.segment not in schema_segments:
                    raise ValueError(
                        "sequence contains invalid segment name",
                    )

            if required_schema_segments - {
                sequence.segment for sequence in isolate.sequences
            }:
                raise ValueError(
                    "isolate does not contain all required schema segments",
                )

        sequence_segments = {
            sequence.segment
            for isolate in self.isolates
            for sequence in isolate.sequences
        }

        if schema_segments != sequence_segments:
            raise ValueError("All schema segments must have a corresponding sequence")

        return self
