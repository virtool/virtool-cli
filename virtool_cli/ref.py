from pathlib import Path

import click

import virtool_cli.build
import virtool_cli.divide
import virtool_cli.isolate
import virtool_cli.repair
import virtool_cli.taxid


@click.group("ref")
def ref():
    """Commands related to references"""
    pass


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a database reference directory",
)
@click.option(
    "-o",
    "--output",
    default="reference.json",
    help="the output path for a reference.json file",
)
@click.option("-i", "--indent", is_flag=True)
@click.option(
    "-V",
    "--version",
    default=None,
    type=str,
    help="the version string to include in the reference.json file",
)
def build(src_path, output, indent, version):
    """Build a Virtool reference JSON file from a reference directory."""
    try:
        virtool_cli.build.run(Path(src_path), Path(output), indent, version)
    except (FileNotFoundError, NotADirectoryError):
        click.echo("Not a valid reference directory")


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a input reference.json file",
)
@click.option(
    "-o",
    "--output",
    default="src",
    type=str,
    help="the output path for a divided reference directory tree",
)
def divide(src_path, output):
    """Divide a reference.json file from Virtool into a reference directory tree."""
    try:
        if not src_path.endswith(".json"):
            raise TypeError
        virtool_cli.divide.run(Path(src_path), Path(output))
    except (TypeError, FileNotFoundError):
        click.echo(
            "Specified reference file either does not exist or is not a proper JSON file"
        )


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to the input reference.json file",
)
@click.option("-f", "--force_update", is_flag=True)
def taxid(src_path, force_update):
    """Fetch taxid for all OTU in given reference directory."""
    try:
        virtool_cli.taxid.run(Path(src_path), force_update)
    except (FileNotFoundError, NotADirectoryError):
        click.echo("Not a valid reference directory")


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a reference directory",
)
def isolate(src_path):
    """Fetch new isolates for all OTU in a given reference directory."""
    try:
        virtool_cli.isolate.run(Path(src_path))
    except (FileNotFoundError, NotADirectoryError):
        click.echo("Not a valid reference directory")


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a reference directory",
)
def repair(src_path):
    """Fix every OTU in a given reference directory."""
    try:
        virtool_cli.repair.run(Path(src_path))
    except (FileNotFoundError, NotADirectoryError):
        click.echo("Not a valid reference directory")


if __name__ == "__main__":
    ref()
