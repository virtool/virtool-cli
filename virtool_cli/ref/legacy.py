"""Code for the old reference format, not based on event sourcing."""

import json
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Any

import click

from virtool_cli.utils.id_generator import (
    generate_random_alphanumeric,
)
from virtool_cli.utils.reference import generate_otu_dirname, is_v1


@dataclass
class RepoSequence:
    """Represents a sequence in a Virtool reference repository."""

    id: str
    """The sequence id."""

    accession: str
    """The sequence accession."""

    definition: str
    """The sequence definition."""

    host: str
    """The sequence host."""

    isolate: "RepoIsolate"
    """The parent isolate."""

    sequence: str
    """The sequence."""

    segment: str
    """The sequence segment."""


@dataclass
class RepoIsolate:
    """Represents an isolate in a Virtool reference repository."""

    def __init__(
        self,
        id_: str,
        default: bool,
        otu: "RepoOTU",
        path: Path,
        source_name: str,
        source_type: str,
    ):
        self.id = id_
        """The isolate id."""

        self.default = default
        """Indicates whether the isolate is the Virtool default for the OTU."""

        self.otu = otu
        """The isolate's parent OTU."""

        self.path = path
        """The path to the isolate directory."""

        self.source_name = source_name
        """The isolate's source name."""

        self.source_type = source_type
        """The isolate's source type."""

        self.sequences: list[RepoSequence] = []
        """A list of child sequences."""

        for sequence_path in (
            p for p in self.path.iterdir() if p.name != "isolate.json"
        ):
            with open(sequence_path) as f:
                sequence_data = json.load(f)
                sequence = RepoSequence(
                    sequence_data["_id"],
                    sequence_data["accession"],
                    sequence_data["definition"],
                    sequence_data["host"],
                    self,
                    sequence_data["sequence"],
                    sequence_data["segment"],
                )
            self.sequences.append(sequence)
        self.sequences.sort(key=lambda s: s.accession)

    @property
    def name(self) -> str:
        """The isolate name."""
        return f"{self.source_type} {self.source_name}"

    def to_dict(self) -> dict[str, Any]:
        sequences = []

        for sequence in self.sequences:
            sequence_dict = asdict(sequence)
            sequence_dict.pop("isolate")

        return {
            "id": self.id,
            "default": self.default,
            "sequences": sequences,
            "source_name": self.source_name,
            "source_type": self.source_type,
        }

    def add_sequence(
        self,
        accession: str,
        definition: str,
        host: str,
        segment: str,
        sequence: str,
    ) -> RepoSequence | None:
        """Add a sequence to the isolate."""
        if accession in self.otu.exclusions:
            click.echo(f"accession {accession} is in exclusions list", err=True)
            return None

        existing_accessions = {
            a.split(".")[0]
            for a in self.otu.repo.maps.sequence_id_to_accession.values()
        }

        if accession.split(".")[0] in existing_accessions:
            click.echo(f"accession {accession} already exists in repo", err=True)
            return None

        sequence_id = generate_random_alphanumeric(
            length=8,
            excluded=set(self.otu.repo.maps.sequence_id_to_path.keys()),
        )

        sequence_path = self.path / f"{sequence_id}.json"

        with open(sequence_path, "w") as f:
            json.dump(
                {
                    "_id": sequence_id,
                    "accession": accession,
                    "definition": definition,
                    "host": host,
                    "segment": segment,
                    "sequence": sequence,
                },
                f,
                indent=4,
                sort_keys=True,
            )

        sequence = RepoSequence(
            sequence_id,
            accession,
            definition,
            host,
            self,
            sequence,
            segment,
        )
        self.otu.repo.maps.sequence_id_to_otu_id[sequence_id] = self.otu.id
        self.otu.repo.maps.sequence_id_to_path[sequence_id] = sequence_path

        self.sequences.append(sequence)

        return sequence


@dataclass
class RepoOTU:
    """Represents an OTU in a Virtool reference repository."""

    def __init__(self, id_: str, path: Path, repo: "Repo"):
        self.id = id_
        """The OTU id."""

        self.path = path
        """The path to the OTU directory."""

        self.repo = repo
        """The OTU's parent repository."""

        self.exclusions = []
        """
        A list of accessions that have been blacklisted for this OTU.

        They should not be fetched again or nested in this OTU.
        """

        self.isolates = []
        """A list of child isolates."""

        self._isolate_name_id_map: dict[tuple[str, str], str] = {}
        """Maps isolate source type-name combos to isolate ids."""

        self.abbreviation = ""
        """The OTU abbreviation (eg. TMV)."""

        self.name = ""
        """The OTU name (eg. Tobacco mosaic virus)."""

        self.schema = []

        self.taxid: int = -1
        """The OTU taxonomy ID."""

        for isolate_path in [p for p in self.path.iterdir() if p.is_dir()]:
            with open(isolate_path / "isolate.json") as f:
                isolate_data = json.load(f)

            isolate = RepoIsolate(
                isolate_data["id"],
                isolate_data["default"],
                self,
                isolate_path,
                isolate_data["source_name"],
                isolate_data["source_type"],
            )

            self.isolates.append(isolate)

            self._isolate_name_id_map[(isolate.source_type, isolate.source_name)] = (
                isolate.id
            )

            with open(self.path / "otu.json") as f:
                data = json.load(f)

                if self.id != data["_id"]:
                    raise ValueError("otu_id does not match otu.json _id")

                self.abbreviation = data["abbreviation"]
                """The OTU abbreviation (eg. TMV)."""

                self.name = data["name"]
                """The OTU name (eg. Tobacco mosaic virus)."""

                self.schema = data["schema"]

                self.taxid: int = int(data["taxid"])
                """The OTU taxonomy ID."""

            with open(self.path / "exclusions.json") as f:
                self.exclusions = json.load(f)

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "abbreviation": self.abbreviation,
            "name": self.name,
            "schema": self.schema,
            "taxid": self.taxid,
            "isolates": [i.to_dict() for i in self.isolates],
            "exclusions": self.exclusions,
        }

    @property
    def blocked_accessions(self):
        existing_accessions = []

        for isolate in self.isolates:
            for sequence in isolate.sequences:
                existing_accessions.append(sequence.accession)

        return sorted(existing_accessions + self.exclusions)

    def add_isolate(self, source_type: str, source_name: str):
        """Add a new isolate to the OTU."""
        isolate_id = generate_random_alphanumeric(
            length=8,
            excluded=set(self.repo.maps.isolate_id_to_otu_id.keys()),
        )

        isolate_path = self.path / isolate_id
        isolate_path.mkdir()

        with open(isolate_path / "isolate.json", "w") as f:
            json.dump(
                {
                    "id": isolate_id,
                    "default": False,
                    "source_name": source_name,
                    "source_type": source_type,
                },
                f,
                indent=4,
                sort_keys=True,
            )

        isolate = RepoIsolate(
            isolate_id,
            False,
            self,
            isolate_path,
            source_name,
            source_type,
        )

        self.isolates.append(isolate)

        self._isolate_name_id_map[(source_type, source_name)] = isolate_id

        return isolate

    def get_isolate_by_id(self, isolate_id: str) -> RepoIsolate:
        """Get an isolate by its id."""
        for isolate in self.isolates:
            if isolate.id == isolate_id:
                return isolate

        raise ValueError(f"isolate_id {isolate_id} not found")

    def get_isolate_by_name(
        self,
        source_type: str,
        source_name: str,
    ) -> RepoIsolate | None:
        """Get an isolate by its source type and name."""
        try:
            isolate_id = self._isolate_name_id_map[(source_type, source_name)]
        except KeyError:
            return None

        return self.get_isolate_by_id(isolate_id)

    def update(
        self,
        abbreviation: str | None = None,
        exclusions: list[str] | None = None,
        name: str | None = None,
    ):
        json_path = self.path / "otu.json"

        with open(json_path) as f:
            otu = json.load(f)

            if abbreviation is not None:
                otu["abbreviation"] = abbreviation

            if name is not None:
                otu["name"] = name

        with open(json_path, "w") as f:
            json.dump(otu, f, indent=4, sort_keys=True)

        with open(self.path / "exclusions.json", "w") as f:
            json.dump(exclusions, f)


class RepoMaps:
    def __init__(self, path: Path):
        self.isolate_id_to_otu_id: dict[str, str] = {}
        """Maps isolate ids to parent otu ids."""

        self.otu_id_to_path: dict[str, Path] = {}
        """Maps otu ids to their respective paths."""

        self.sequence_id_to_accession: dict[str, str] = {}
        """Maps sequence ids to their respective accessions."""

        self.sequence_id_to_otu_id: dict[str, str] = {}
        """Maps sequence ids to parent otu ids."""

        self.sequence_id_to_path: dict[str, Path] = {}
        """Maps sequence ids to their respective paths."""

        self.taxid_to_otu_id: dict[int, str] = {}
        """Maps taxids to otu ids."""

        for otu_path in (p for p in (path / "src").iterdir() if p.is_dir()):
            if not otu_path.is_dir():
                continue

            name, otu_id = otu_path.name.split("--")

            self.otu_id_to_path[otu_id] = otu_path

            with open(otu_path / "otu.json") as f:
                taxid = int(json.load(f)["taxid"])

            self.taxid_to_otu_id[taxid] = otu_id

            for isolate_path in (p for p in otu_path.iterdir() if p.is_dir()):
                self.isolate_id_to_otu_id[isolate_path.stem] = otu_id

                for sequence_path in [
                    p for p in isolate_path.iterdir() if p.name != "isolate.json"
                ]:
                    sequence_id = sequence_path.stem

                    with open(sequence_path) as f:
                        self.sequence_id_to_accession[sequence_id] = json.load(f)[
                            "accession"
                        ]

                    self.sequence_id_to_otu_id[sequence_id] = otu_id
                    self.sequence_id_to_path[sequence_id] = sequence_path


class Repo:
    def __init__(self, path: Path):
        self.path = path
        """The path to the repo directory."""

        if is_v1(self.path / "src"):
            raise ValueError("Repository is a deprecated v1 reference.")

        with open(self.path / "src" / "meta.json") as f:
            data = json.load(f)

            self.data_type = data["data_type"]
            """The sequence data type of the repo (eg. genome)."""

            self.organism = data["organism"]
            """The type of organism organized in the reference (eg. virus)."""

        self.maps = RepoMaps(self.path)

    @property
    def src_path(self) -> Path:
        """The path to the repo src directory."""
        return self.path / "src"

    def get_otu_path(self, otu_id: str) -> Path:
        """Get the path to the otu with the given ``otu_id``."""
        try:
            return self.maps.otu_id_to_path[otu_id]
        except KeyError:
            raise ValueError(f"No OTU with ID {otu_id}")

    def get_sequence_path(self, sequence_id: str) -> Path:
        """Get the path to the sequence with the given ``sequence_id``."""
        try:
            return self.maps.sequence_id_to_path[sequence_id]
        except KeyError:
            raise ValueError(f"sequence_id {sequence_id} not found")

    def get_otu_by_id(self, otu_id: str):
        """Get an OTU by its ID."""
        try:
            path = self.get_otu_path(otu_id)
        except ValueError:
            raise ValueError(f"No OTU with ID {otu_id}")

        return RepoOTU(
            otu_id,
            path,
            self,
        )

    def get_otu_by_taxid(self, taxid: int) -> RepoOTU:
        try:
            otu_id = self.maps.taxid_to_otu_id[taxid]
        except KeyError:
            raise ValueError(f"No OTU with taxid {taxid}")

        return self.get_otu_by_id(otu_id)

    def create_otu(self, name: str, taxid: int, abbreviation: str = "") -> RepoOTU:
        """Create a new OTU."""
        otu_id = generate_random_alphanumeric(
            length=8,
            excluded=list(self.maps.otu_id_to_path.keys()),
        )

        otu_path = self.src_path / generate_otu_dirname(name, otu_id=otu_id)

        self.maps.otu_id_to_path[otu_id] = otu_path
        self.maps.taxid_to_otu_id[taxid] = otu_id

        try:
            otu_path.mkdir()

            with open(otu_path / "otu.json", "w") as f:
                json.dump(
                    {
                        "_id": otu_id,
                        "abbreviation": abbreviation,
                        "name": name,
                        "schema": [],
                        "taxid": taxid,
                    },
                    f,
                    indent=4,
                    sort_keys=True,
                )

            with open(otu_path / "exclusions.json", "w") as f:
                json.dump([], f, indent=4, sort_keys=True)
        except Exception as e:
            try:
                otu_path.rmdir()
            except FileNotFoundError:
                pass

            try:
                del self.maps.otu_id_to_path[otu_id]
            except KeyError:
                pass

            try:
                del self.maps.taxid_to_otu_id[taxid]
            except KeyError:
                pass

            raise e

        return self.get_otu_by_id(otu_id)

    def iter_otus(self):
        """Iterate over all OTUs."""
        for otu_id in self.maps.otu_id_to_path:
            yield self.get_otu_by_id(otu_id)
