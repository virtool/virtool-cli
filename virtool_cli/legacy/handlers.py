import json

from virtool_cli.legacy.utils import ErrorHandledResult, HandleErrorContext
from virtool_cli.ref.otu import extract_isolate_source


def handle_enum(ctx: HandleErrorContext) -> ErrorHandledResult:
    """Handle a Pydantic ``enum`` error.

    If the field is ``source_type`` and fixing is enabled, the accessions associated
    with the isolate will be fetched from NCBI and the source name and source type will
    be derived and updated in the isolate.

    If no source type can be found or derived, the source_type will be set to "genbank"
    and the source_name will be set to the earliest sorted accession for the associated
    sequences.

    :param ctx: the error context
    :return: the error handling result

    """
    expected = ctx.error["ctx"]["expected"]
    field = ctx.error["loc"][-1]
    value = ctx.error["input"]

    message = (
        f"[bold]{field}[/bold] must be one of [u]{expected}[/u], but is "
        f"[u]{json.dumps(value).replace('"', "'")}[/u]."
    )

    if field == "source_type" and ctx.fix:
        isolate_index = ctx.error["loc"][1]
        isolate = ctx.otu["isolates"][isolate_index]

        genbank_records = ctx.ncbi_client.fetch_genbank_records(
            [s["accession"] for s in isolate["sequences"]],
        )

        try:
            source = extract_isolate_source(genbank_records)
        except ValueError:
            return ErrorHandledResult(
                f"{message}. no isolate source could be found on genbank.",
                False,
            )

        ctx.update_isolate(
            isolate_index,
            {"source_name": source.name, "source_type": source.type.lower()},
        )

        return ErrorHandledResult(message, True)

    return ErrorHandledResult(message)


def handle_int_type(ctx: HandleErrorContext) -> ErrorHandledResult:
    """Handle a Pydantic ``int_type`` error.

    If the field is ``taxid`` and fixing is enabled, the taxid will be fetched from NCBI
    using the OTU name and updated in the OTU.

    :param ctx: the error context
    :return: the error handling result
    """
    field = ctx.error["loc"][-1]
    value = ctx.error["input"]

    result = ErrorHandledResult(
        f"[bold]{field}[/bold] must be an integer. invalid value is "
        f"[u]{json.dumps(value)}[/u].",
    )

    if field == "taxid" and ctx.fix:
        taxid = ctx.ncbi_client.fetch_taxonomy_id_by_name(ctx.otu["name"])

        if taxid is None:
            return ErrorHandledResult(
                f"[bold]taxid[/bold] was not an integer and a taxid could not be "
                f"found for [u]{ctx.otu['name']}[/u] on ncbi.",
            )

        ctx.update_otu({"taxid": int(taxid)})

        result.fixed = True

        return result

    return result


def handle_missing(ctx: HandleErrorContext) -> ErrorHandledResult:
    """Handle a Pydantic ``missing`` error.

    If the field is ``taxid`` and fixing is enabled, the taxid will be fetched from NCBI
    using the OTU name and updated in the OTU.

    :param ctx: the error context
    :return: the error handling result

    """
    field = ctx.error["loc"][-1]

    message = f"[bold]{field}[/bold] is missing"

    if field == "taxid" and ctx.fix:
        taxid = ctx.ncbi_client.fetch_taxonomy_id_by_name(ctx.otu["name"])

        if taxid is None:
            return ErrorHandledResult(
                f"{message} and a taxid could not be found for [u]{ctx.otu['name']}[/u]"
                f" on ncbi.",
            )

        ctx.update_otu({"taxid": taxid})

        return ErrorHandledResult(message, fixed=True)

    return ErrorHandledResult(message)


def handle_str_type(ctx: HandleErrorContext) -> ErrorHandledResult:
    """Handle a Pydantic ``str_type`` error.

    If the field is ``accession`` and fixing is enabled, the accession will be fetched
    from NCBI using the OTU name and updated in the OTU.

    If the field is ``host`` and fixing is enabled, the host will be set to an empty
    string.

    :param ctx: the error context
    :return: the error handling result

    """
    value = ctx.error["input"]
    field = ctx.error["loc"][-1]

    if field == "host":
        return ErrorHandledResult(
            f"[bold]{field}[/bold] must be an string. invalid value is "
            f'[u]{json.dumps(value)}[/u]. set to [u]""[/u].',
            fixed=True,
        )

    return ErrorHandledResult(
        f"[bold]{field}[/bold] must be an string. invalid value is "
        f"[u]{json.dumps(value)}[/u].",
    )


def handle_string_pattern_mismatch(ctx: HandleErrorContext) -> ErrorHandledResult:
    """Handle a Pydantic ``string_pattern_mismatch`` error.

    These errors occur when an accession or genomic sequence do not match defined regex
    patterns.

    If the field is ``accession`` and the value does not contain a version suffix, the
    accession will be fetched from NCBI and the versioned accession will be attached
    to the sequence. Any other fields associated with the sequence will also be updated.

    If the field is ``sequence``, the error message will be updated to include the
    incorrect sequence characters.

    :param ctx: the error context
    :return: the error handling result
    """
    field = ctx.error["loc"][-1]
    pattern = ctx.error["ctx"]["pattern"]
    value = ctx.error["input"]

    if field == "sequence":
        invalid_chars = [
            (i, char) for i, char in enumerate(value) if char not in "ATCGRYKMSWBDHVN"
        ]

        mismatches = ",".join([f"{char}@{i}" for i, char in invalid_chars])

        return ErrorHandledResult(
            f"[bold]{field}[/bold] contains invalid characters. invalid "
            f"characters are [u]{mismatches}[/u].",
        )

    if field == "accession" and "." not in value:
        message = (
            f"[bold]accession[/bold] value [u]{value}[/u] does not contain a "
            f"[u].[/u] and needs a valid version suffix."
        )

        if not ctx.fix:
            return ErrorHandledResult(message)

        genbank_records = ctx.ncbi_client.fetch_genbank_records(
            [value],
        )

        if not genbank_records:
            return ErrorHandledResult(
                message + " couldn't fix because no matching genbank record found.",
            )

        genbank_record = genbank_records[0]

        isolate_index = ctx.error["loc"][1]
        sequence_index = ctx.error["loc"][3]

        ctx.update_sequence(
            isolate_index,
            sequence_index,
            {
                "accession": genbank_record.accession_version,
                "definition": genbank_record.definition,
                "sequence": genbank_record.sequence,
            },
        )

        return ErrorHandledResult(message, fixed=True)

    return ErrorHandledResult(
        f"[bold]{field}[/bold] does not match the required regex pattern "
        f"[u]{pattern}[/u].",
    )


def handle_string_too_short(ctx: HandleErrorContext) -> ErrorHandledResult:
    """Handle a Pydantic ``string_too_short`` error.

    If the field is ``source_name`` and fixing is enabled, the accessions associated
    with the isolate will be fetched from NCBI and the source name and source type will
    be derived and updated in the isolate.

    If no source type can be found or derived, an informative error message will be
    returned.

    :param ctx: the error context
    :return: the error handling result
    """
    field = ctx.error["loc"][-1]
    input_ = ctx.error["input"]
    min_length = ctx.error["ctx"]["min_length"]

    result = ErrorHandledResult(
        f"[bold]{field}[/bold] must be a string with a minimum length of "
        f"{min_length}. it currently has a length of [u]{len(input_)}[/u].",
    )

    if field == "source_name" and ctx.fix:
        isolate_index = ctx.error["loc"][1]
        isolate = ctx.otu["isolates"][isolate_index]

        genbank_records = ctx.ncbi_client.fetch_genbank_records(
            [s["accession"] for s in isolate["sequences"]],
        )

        if not genbank_records:
            return ErrorHandledResult(
                f"[bold]source_name[/bold] must be a string with a minimum length of "
                f"{min_length}. it currently has a length of [u]{len(input_)}[/u]. we "
                "could not find a source name on genbank.",
            )

        source = extract_isolate_source(genbank_records)

        ctx.update_isolate(
            isolate_index,
            {"source_name": source.name, "source_type": source.type.lower()},
        )

        result.fixed = True

    return result


def handle_too_short(ctx: HandleErrorContext) -> ErrorHandledResult:
    """Handle a Pydantic ``too_short`` error.

    Return a result object with an informative, human-readable error message.

    :param ctx: the error context
    :return: the error handling result
    """
    actual_length = ctx.error["ctx"]["actual_length"]
    field = ctx.error["loc"][-1]
    field_type = ctx.error["ctx"]["field_type"]
    min_length = ctx.error["ctx"]["min_length"]

    return ErrorHandledResult(
        f"[bold]{field}[/bold] must be a {field_type} with a minimum length of "
        f"{min_length}. it currently has a length of [u]{actual_length}[/u].",
    )


def handle_value_error(ctx: HandleErrorContext) -> ErrorHandledResult:
    """Handle a ``ValueError`` by returning a result object with a human-readable error
    message.

    :param ctx: the error context
    :return: the error handling result
    """
    return ErrorHandledResult(str(ctx.error["ctx"]["error"]))
