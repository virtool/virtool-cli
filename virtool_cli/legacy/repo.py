"""Repo level checks for legacy repositories."""

from pathlib import Path

from virtool_cli.legacy.utils import iter_legacy_otus
from virtool_cli.utils.console import console


def check_unique_accessions(path: Path):
    """Validate that all accession fields are unique."""
    accessions = set()

    for otu in iter_legacy_otus(path / "src"):
        for isolate in otu["isolates"]:
            for sequence in isolate["sequences"]:
                if accession := sequence.get("accession"):
                    if accession in accessions:
                        console.log(f"Found non-unique accession: {accession}")
                    else:
                        accessions.add(accession)


def check_unique_ids(path: Path):
    """Validate that resource IDs are unique.

    * All OTU _id fields are unique.
    * All isolate id fields are unique.
    * All sequence _id fields are unique.
    """
    otu_ids = []
    isolate_ids = []
    sequence_ids = []

    for otu in iter_legacy_otus(path / "src"):
        otu_ids.append(otu["_id"])

        for isolate in otu["isolates"]:
            isolate_ids.append(isolate["id"])

            for sequence in isolate["sequences"]:
                sequence_ids.append(sequence["_id"])

    if len(otu_ids) != len(set(otu_ids)):
        console.log("Found non-unique OTU ids")

    if len(isolate_ids) != len(set(isolate_ids)):
        console.log("Found non-unique isolate ids")

    if len(sequence_ids) != len(set(sequence_ids)):
        console.log("Found non-unique sequence ids")


def check_unique_otu_abbreviations_and_names(path: Path):
    """Validate that OTU abbreviations and names are unique."""
    abbreviations = set()
    names = set()

    for otu in iter_legacy_otus(path / "src"):
        abbreviation = otu["abbreviation"]

        if abbreviation:
            if abbreviation in abbreviations:
                console.log(f"Found non-unique OTU abbreviation: {abbreviation}")
            else:
                abbreviations.add(abbreviation)

        name = otu["name"].lower()

        if name in names:
            console.log(f"Found non-unique OTU name: {name}")
        else:
            names.add(name)
