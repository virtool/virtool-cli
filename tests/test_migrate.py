import shutil
import subprocess
from pathlib import Path

from syrupy import SnapshotAssertion

from virtool_cli.ref.legacy import Repo


def test_migrate(test_files_path: Path, snapshot: SnapshotAssertion, tmp_path: Path):
    """Test that the generated catalog creates all the same filenames as
    the control catalog
    """
    path = tmp_path / "v1"
    path.mkdir()

    migrate_path = path / "src"
    shutil.copytree(test_files_path / "src_v1", migrate_path)

    subprocess.call(["virtool", "ref", "migrate", "-src", str(migrate_path)])

    # Make sure the repo can be loaded.
    repo = Repo(path)

    # Make sure data from meta.json is available.
    assert repo.data_type == "genome"
    assert repo.organism == "virus"

    # Make sure the OTU directories are renamed.
    assert {
        otu_id: path.name for otu_id, path in repo.maps.otu_id_to_path.items()
    } == snapshot

    # Make sure exclusions.json files are created.
    assert all(
        (otu_path / "exclusions.json").exists()
        for otu_path in repo.maps.otu_id_to_path.values()
    )
