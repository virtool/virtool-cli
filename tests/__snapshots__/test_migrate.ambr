# serializer version: 1
# name: test_migrate
  dict({
    '0925e7b6': 'vanilla_distortion_mosaic_virus--0925e7b6',
    '0bfdb8bc': 'wasabi_mottle_virus--0bfdb8bc',
    '2031cb7f': 'impatiens_necrotic_spot_virus--2031cb7f',
    '34f59aff': 'fig_mosaic_virus--34f59aff',
    '3962f6ec': 'jacquemontia_mosaic_yucatan_virus--3962f6ec',
    '407d8a21': 'kalanchoe_latent_virus--407d8a21',
    '5d877144': 'yacon_necrotic_mottle_virus--5d877144',
    '6bc81904': 'saccharum_streak_virus--6bc81904',
    '71d15b86': 'ugandan_cassava_brown_streak_virus--71d15b86',
    '7d0ffa77': 'east_asian_passiflora_virus--7d0ffa77',
    '9da7c10d': 'radish_leaf_curl_betasatellite--9da7c10d',
    'a89b6529': 'habenaria_mosaic_virus--a89b6529',
    'ab6bd88a': 'dahlia_latent_viroid--ab6bd88a',
    'ae0f2a35': 'nanovirus_like_particle--ae0f2a35',
    'c55afad6': 'bamboo_mosaic_virus--c55afad6',
    'c93ec9a9': 'abaca_bunchy_top_virus--c93ec9a9',
    'd226290f': 'cabbage_leaf_curl_jamaica_virus--d226290f',
    'd86d0412': 'tall_oatgrass_mosaic_virus--d86d0412',
    'dd21fd8f': 'pagoda_yellow_mosaic_associated_virus--dd21fd8f',
    'e1dfb3ef': 'oat_blue_dwarf_virus--e1dfb3ef',
    'e5b6dd57': 'zantedeschia_mild_mosaic_virus--e5b6dd57',
    'edfaaf2b': 'macroptilium_golden_mosaic_virus--edfaaf2b',
    'f8a56910': 'gaillardia_latent_virus--f8a56910',
    'fbd9391c': 'lamium_leaf_distortion_virus--fbd9391c',
  })
# ---
